### PR TITLE
[mlir][llvm] Handle invoke op branching to block with its result as an argument

### PR DESCRIPTION
**Pull request by victor-eds**
In LLVM, having an invoke instruction branching to a block with a phi
node receiving the invoke instruction result as an argument is
perfectly legal. However, the translation of this construct to MLIR
would result in an invoke with its result being used as a block
argument to a successor, i.e., the operation result would be used in
its definition.

In order to fix this issue due to different IR structures (phi nodes
vs block arguments), this construct is interpreted with an llvm.invoke
operation branching to a dummy block having a single llvm.br operation
passing the required block arguments (including the result of the
llvm.invoke operation) to the actual successor block.

Differential Revision: https://reviews.llvm.org/D148313

### CODE REVIEW

**Comment by whitneywhtsang** [comment]
> This PR should not be merged, as we will take the "get reviews upstream first" approach.

Thinking if we should use a label (maybe repurpose an existing one e.g., `upstream`) to indicate PR we want to go directly upstream.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp [hunk 1]
```diff
@@ -1293,28 +1293,78 @@ LogicalResult ModuleImport::convertInstruction(llvm::Instruction *inst) {
...
+    std::size_t i = 0;
+    std::optional<std::size_t> invokeBlockArgPos;
+    for (auto inst = invokeInst->getNormalDest()->begin();
+         !invokeBlockArgPos && isa<llvm::PHINode>(inst); ++inst, ++i) {
+      auto *phiInst = cast<llvm::PHINode>(&*inst);
+      llvm::Value *value =
+          phiInst->getIncomingValueForBlock(invokeInst->getParent());
+      if (value == invokeInst)
+        invokeBlockArgPos = i;
+    }
```
Comment (Naghasan) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1307]: You do not use the index, so suggesting something like:

```suggestion
    bool invokeResultUsedInPhi = false;
    for (const auto* phiInst : invokeInst->getNormalDest()->phis()) {
      if (phiInst->getIncomingValueForBlock(invokeInst->getParent()) == invokeInst) {
        invokeResultUsedInPhi = true;
        break;
      }
    }
```

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp [hunk 2]
```diff
@@ -1293,28 +1293,78 @@ LogicalResult ModuleImport::convertInstruction(llvm::Instruction *inst) {
...
+                                 normalArgs)))
+      return failure();
+
+    if (invokeBlockArgPos) {
+      // The dummy normal dest block will just host an unconditional branch
+      // instruction to the normal destination block passing the required block
+      // arguments (including the invoke operation's result).
+      OpBuilder::InsertionGuard g(builder);
+      builder.setInsertionPointToStart(directNormalDest);
+      builder.create<LLVM::BrOp>(loc, invokeOp.getResult(0), normalDest);
```
Comment (Naghasan) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1356]: Shouldn't this be

```suggestion
      builder.create<LLVM::BrOp>(loc, normalArgs, normalDest);
```

Looks like you are assuming you can only have 1 argument

File: mlir/test/Target/LLVMIR/Import/basic.ll [hunk 3]
```diff
@@ -87,6 +87,38 @@ define i32 @invokeLandingpad() personality ptr @__gxx_personality_v0 {
...
+  %invoke = invoke i32 @foo2() to label %bb0 unwind label %bb1
+; CHECK: ^bb2:  // pred: ^bb0
+nocall:
+  ; CHECK: llvm.br ^bb4(%0 : i32)
+  br label %bb0
+; CHECK: ^bb3:  // pred: ^bb1
+  ; CHECK: llvm.br ^bb4(%1 : i32)
+; CHECK: ^bb4(%2: i32):  // 2 preds: ^bb2, ^bb3
+bb0:
+  %ret = phi i32 [ 0, %nocall ], [ %invoke, %call ]
```
Comment (Naghasan) [inline mlir/test/Target/LLVMIR/Import/basic.ll:111]: can you test for multiple phis ? something like that (also test operand unrelated to invoke)

```suggestion
  %ret = phi i32 [ 0, %nocall ], [ %invoke, %call ]
  %foo = phi i32 [ 0, %nocall ], [ 1, %call ]
```

**Comment by victor-eds** [comment]
> > This PR should not be merged, as we will take the "get reviews upstream first" approach.
> 
> Thinking if we should use a label (maybe repurpose an existing one e.g., `upstream`) to indicate PR we want to go directly upstream.

I agree. Maybe we can discuss on Monday which to use :+1:

**Comment by victor-eds** [comment]
> I added the existing "upstream" label to indicate this PR has an equivalent upstream PR.

Excellent. I'll mark this PR as a draft and create a patch upstream. Thanks for the reviews, guys!

**Comment by victor-eds** [comment]
This is a mirror of https://reviews.llvm.org/rGbe717f0a49d9591fec6574cb9b186be904a2d604

**Pull request merged**